### DISCUSSION TITLE
Self dependent queries

### DISCUSSION THREAD

**Discussion by cachius** [TanStack/query#5422] [Ideas] [answered]
useQuery nicely supports [dependent queries](https://tanstack.com/query/v5/docs/react/guides/dependent-queries) already.
For polling I often need "self dependent queries" that terminate once a certain result has been received. Please add support for that. The current solution is rather clunky:

```js
const [enabled, setEnabled] = useState(true);
const { isLoading, data } = useQuery(/* ... */, {enabled});

const state = data?.state;
useEffect(() => {
    if (state === 'finished') setEnabled(false);
}, [state]);
```
(via [How to use the react-query result inside the QueryOptions?](https://stackoverflow.com/questions/76264840/how-to-use-the-react-query-result-inside-the-queryoptions))

I think a functional form of QueryOptions would be great:
```js
const { isLoading, data } = useQuery(
  "data-querykey-" + id,
  () => api.getData({ id }),
  (data) => {
    enabled: data?.state !== "finished",
    refetchInterval: 3000,
  }
);
```
with data being `undefined` in the first iteration.

**Comment by TkDodo** [comment, marked as answer]
refetchInterval can be a function to support exactly this usecase
